src/Avalonia.Base/Data/Optional.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace Avalonia.Data
+{
+    /// <summary>
+    /// An optional typed value.
+    /// </summary>
+    /// <typeparam name="T">The value type.</typeparam>
+    /// <remarks>
+    /// This struct is similar to <see cref="Nullable{T}"/> except it also accepts reference types:
+    /// note that null is a valid value for reference types. It is also similar to
+    /// <see cref="BindingValue{T}"/> but has only two states: "value present" and "value missing".
+    /// 
+    /// To create a new optional value you can:
+    /// 
+    /// - For a simple value, call the <see cref="Optional{T}"/> constructor or use an implicit
+    ///   conversion from <typeparamref name="T"/>
+    /// - For an missing value, use <see cref="Empty"/> or simply `default`
+    /// </remarks>
+    public readonly struct Optional<T> : IEquatable<Optional<T>>
+    {
+        private readonly T _value;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Optional{T}"/> struct with value.
+        /// </summary>
+        /// <param name="value">The value.</param>
+        public Optional(T value)
+        {
+            _value = value;
+            HasValue = true;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether a value is present.
+        /// </summary>
+        public bool HasValue { get; }
+
+        /// <summary>
+        /// Gets the value.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// <see cref="HasValue"/> is false.
+        /// </exception>
+        public T Value => HasValue ? _value : throw new InvalidOperationException("Optional has no value.");
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj) => obj is Optional<T> o && this == o;
+
+        /// <inheritdoc/>
+        public bool Equals(Optional<T> other) => this == other;
+
+        /// <inheritdoc/>
+        public override int GetHashCode() => HasValue ? Value!.GetHashCode() : 0;
+
+        /// <summary>
+        /// Casts the value (if any) to an <see cref="object"/>.
+        /// </summary>
+        /// <returns>The cast optional value.</returns>
+        public Optional<object> ToObject() => HasValue ? new Optional<object>(Value) : default;
+
+        /// <inheritdoc/>
+        public override string ToString() => HasValue ? Value?.ToString() ?? "(null)" : "(empty)";
+
+        /// <summary>
+        /// Gets the value if present, otherwise a default value.
+        /// </summary>
+        /// <param name="defaultValue">The default value.</param>
+        /// <returns>The value.</returns>
+        public T ValueOrDefault(T defaultValue = default) => HasValue ? Value : defaultValue;
```

Comment:
Perf issue that I spotted when profiling your code - you are double checking `HasValue` in many places. This function ends up checking `HasValue` twice and bloats this function so it won't be inlined in many cases.

Reply:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Nullable.cs

Reply:
RE: double checking `HasValue`, can you explain more? Do you mean that it's checked in `ValueOrDefault` and then in `Value`? The `Nullable` code that you linked to appears to do the same, is it different because it checks a field and not a property?

Also I should rename this method `GetValueOrDefault` to match `Nullable`.

Reply:
Nullable code does it a bit differently:
```csharp
public T GetValueOrDefault(T defaultValue) =>
            hasValue ? value : defaultValue;
```

In most of the code it tries to avoid using `Value` if we know that it is set (by checking `hasValue`), otherwise like in your code we end up doing:
- check if value is set
    - it is set
        - check if value is set (entering potential EH scope)
        - return value
    - it is not set
        - return default value

Reply:
Ah yeah, apologies, I missed the fact that `Nullable` accesses the field directly in `GetValueOrDefault`.

Reply:
@MarchingCube hopefully fixed this in recent commits.